src/posix/platform/resolver.cpp

```diff
@@ -0,0 +1,293 @@
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "resolver.hpp"
+
+#include "platform-posix.h"
+
+#include <openthread/logging.h>
+#include <openthread/message.h>
+#include <openthread/udp.h>
+#include <openthread/platform/dns.h>
+
+#include "common/code_utils.hpp"
+
+#include <arpa/inet.h>
+#include <arpa/nameser.h>
+#include <cassert>
+#include <netinet/in.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+
+#include <fstream>
+#include <string>
+
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+
+namespace {
+constexpr char kResolveConfLocation[] = "/etc/resolv.conf";
+constexpr char kNameserverItem[]      = "nameserver";
+} // namespace
+
+extern ot::Posix::Resolver gResolver;
+
+namespace ot {
+namespace Posix {
+
+void Resolver::Init(void)
+{
+    memset(mUpstreamTransaction, 0, sizeof(mUpstreamTransaction));
+
+    LoadDnsServerListFromConf();
+}
+
+void Resolver::LoadDnsServerListFromConf(void)
+{
+    std::string   line;
+    std::ifstream fp;
+
+    fp.open(kResolveConfLocation);
+    if (fp.bad())
+    {
+        otLogCritPlat("Cannot read %s for domain name servers, default to 127.0.0.1", kResolveConfLocation);
+        mUpstreamDnsServerCount = 1;
+        assert(inet_pton(AF_INET, "127.0.0.1", &mUpstreamDnsServerList[0]) == 1);
+        ExitNow();
+    }
+
+    while (std::getline(fp, line) && mUpstreamDnsServerCount < kMaxUpstreamServerCount)
+    {
+        if (line.find(kNameserverItem, 0) == 0)
+        {
+            in_addr_t addr;
+
+            if (inet_pton(AF_INET, &line.c_str()[sizeof(kNameserverItem)], &addr) == 1)
+            {
+                otLogCritPlat("Got nameserver #%d: %s", mUpstreamDnsServerCount,
+                              &line.c_str()[sizeof(kNameserverItem)]);
+                mUpstreamDnsServerList[mUpstreamDnsServerCount] = addr;
+                mUpstreamDnsServerCount++;
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void Resolver::Query(otPlatDnsUpstreamQuery *aTxn, const otMessage *aQuery)
+{
+    char        packet[kMaxDnsMessageSize];
+    otError     error  = OT_ERROR_NONE;
+    uint16_t    length = otMessageGetLength(aQuery);
+    sockaddr_in serverAddr;
+
+    Transaction *txn;
+
+    VerifyOrExit(length <= kMaxDnsMessageSize, error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(otMessageRead(aQuery, 0, &packet, sizeof(packet)) == length, error = OT_ERROR_NO_BUFS);
+
+    txn = AllocateTransaction(aTxn);
+    VerifyOrExit(txn != nullptr, error = OT_ERROR_NO_BUFS);
+
+    serverAddr.sin_family = AF_INET;
+    serverAddr.sin_port   = htons(53);
+    for (int i = 0; i < mUpstreamDnsServerCount; i++)
+    {
+        serverAddr.sin_addr.s_addr = mUpstreamDnsServerList[i];
+        VerifyOrExit(sendto(txn->mUdpFd, packet, length, 0, (struct sockaddr *)&serverAddr, sizeof(serverAddr)),
+                     error = OT_ERROR_NO_ROUTE);
```

Comment:
May be fine for now.

But thinking if this can block?  If so I guess we may need add it in `UpdateFtSet()` in `WriteFdSt`?

Reply:
Added `MSG_DONTWAIT`

I think we can just act as a packet loss and let the client retry? Consider the fact that we are not holding the ownership. Preallocating buffers or doing another copy here is not a good idea for me?

Also, blocking on sendto for UDP means the device is too busy to send this message, thus throw away the packet makes sense to me.

Reply:
Sounds good to me. I think it was fine to do this or even keep it as before.

If we want to change it to use `UpdateFdSet()`, I feel will become complex and only to handle a rather unlikely edge-case.